File: 2 - Work/Phase 5 The Hawaii Micro-Case Study/Bulk Tests/Julia/Step5_Gap_Analysis.jl
# Purpose: Merge Step 2 TMKs with the Honolulu cadastral CSV to extract TMK
#          zones and map each zone to its official Oahu geographic district.
# Inputs:  Bulk Tests/Julia/Target_Golf_Parcels_List.csv               (Step 2 output)
#          00 - Data Sources/Honolulu/All_Parcels_-4613852522541990741.csv
# Outputs: Bulk Tests/Julia/Phase5_Step5_Geographic_Breakdown.csv


# === 1. USING ===

using DataFrames
using CSV
using Printf


# === 2. GLOBALS & PATHS ===

const SCRIPT_DIR        = @__DIR__
const WORK_DIR          = normpath(joinpath(@__DIR__, "..", "..", ".."))
const HONOLULU_DATA_DIR = joinpath(WORK_DIR, "00 - Data Sources", "Honolulu")
const TMK_LIST_PATH     = joinpath(SCRIPT_DIR, "Target_Golf_Parcels_List.csv")
const TAX_CSV_PATH      = joinpath(HONOLULU_DATA_DIR, "All_Parcels_-4613852522541990741.csv")
const OUT_CSV           = joinpath(SCRIPT_DIR, "Phase5_Step5_Geographic_Breakdown.csv")

const DISTRICT_MAP = Dict(
    "1" => "Honolulu (Urban Core)",
    "2" => "Honolulu (East/Anomalies)",
    "3" => "Honolulu (Anomalies)",
    "4" => "Koolaupoko (Kailua/Kaneohe)",
    "5" => "Koolauloa (North/East)",
    "6" => "Waialua (North Shore)",
    "7" => "Wahiawa (Central)",
    "8" => "Waianae (West)",
    "9" => "Ewa (Kapolei/Pearl City)",
)


# === 3. FUNCTIONS ===

# (none)


# === 4. EXECUTION ===

function main()
    println("\n" * "=" ^ 70)
    println("Phase 5 - Step 5: Geographic Concentration Breakdown")
    println("=" ^ 70 * "\n")

    isfile(TMK_LIST_PATH) || error("[FATAL] TMK list not found. Run Step 2 first.\n  $TMK_LIST_PATH")
    isfile(TAX_CSV_PATH)  || error("[FATAL] Cadastral CSV not found:\n  $TAX_CSV_PATH")

    tmk_df = CSV.read(TMK_LIST_PATH, DataFrame)
    rename!(tmk_df, names(tmk_df)[1] => :TMK)
    tmk_df.TMK_clean = replace.(string.(tmk_df.TMK), r"[^0-9]" => "")

    tax_data    = CSV.read(TAX_CSV_PATH, DataFrame)
    tmk_col_idx = findfirst(c -> occursin(r"^tmk$"i, c), names(tax_data))
    if isnothing(tmk_col_idx)
        println("[FATAL] No TMK column identified in cadastral CSV.")
        println("Available columns: ", names(tax_data))
        error("[FATAL] No TMK column identified.")
    end
    tmk_col = names(tax_data)[tmk_col_idx]
    tax_data.TMK_clean = replace.(string.(tax_data[!, tmk_col]), r"[^0-9]" => "")

    step2_lens = length.(tmk_df.TMK_clean)
    csv_lens   = length.(skipmissing(tax_data.TMK_clean))

    # 8-digit format = Z S PPP QQQ  (3-digit parcel field)
    # 9-digit format = Z S PPP QQQQ (4-digit parcel field, trailing 0 for non-CPR parcels)
    if all(==(8), step2_lens) && all(==(9), csv_lens)
        tmk_df.TMK_clean = tmk_df.TMK_clean .* "0"
    elseif all(==(9), step2_lens) && all(==(8), csv_lens)
        tax_data.TMK_clean = tax_data.TMK_clean .* "0"
    end

    merged_data = innerjoin(tmk_df, tax_data, on = :TMK_clean; makeunique = true)
    # CPR sub-parcel records in the cadastral CSV share a TMK but have null Zone;
    # drop them so only parent parcel records (with zone info) are counted.
    dropmissing!(merged_data, :Zone)

    merged_data.Zone_Code    = string.(merged_data.Zone)
    merged_data.District_Name = map(z -> get(DISTRICT_MAP, z, "Zone $z"), merged_data.Zone_Code)

    geo_summary = combine(
        groupby(merged_data, [:Zone_Code, :District_Name]),
        nrow => :Parcel_Count,
    )
    total_parcels = sum(geo_summary.Parcel_Count)
    geo_summary.Pct_of_Total_Parcels = geo_summary.Parcel_Count ./ total_parcels .* 100
    sort!(geo_summary, :Parcel_Count, rev = true)

    @printf("%-5s %-35s %-15s %-15s\n", "Zone", "Geographic District", "Parcel Count", "% of Parcels")
    println("-" ^ 70)
    for row in eachrow(geo_summary)
        @printf("%-5s %-35s %-15d %.1f%%\n",
                row.Zone_Code, row.District_Name, row.Parcel_Count, row.Pct_of_Total_Parcels)
    end
    println("-" ^ 70)
    @printf("%-5s %-35s %-15d 100.0%%\n", "", "TOTAL", total_parcels)
    println("-" ^ 70)

    CSV.write(OUT_CSV, geo_summary)
    println("\n[+] Geographic Breakdown saved -> $(basename(OUT_CSV))")
    println("\n[DONE] Step 5 Complete. Phase 5 is fully finished!")
end

if abspath(PROGRAM_FILE) == @__FILE__
    main()
end

File: 2 - Work/Phase 5 The Hawaii Micro-Case Study/Bulk Tests/python/Step5_Gap_Analysis.py
# Purpose: Merge Step 2 TMKs with the Honolulu cadastral CSV to extract TMK
#          zones and map each zone to its official Oahu geographic district.
# Inputs:  Bulk Tests/python/Target_Golf_Parcels_List.csv               (Step 2 output)
#          00 - Data Sources/Honolulu/All_Parcels_-4613852522541990741.csv
# Outputs: Bulk Tests/python/Phase5_Step5_Geographic_Breakdown.csv


# === 1. IMPORTS ===

import pathlib
import re
import pandas as pd


# === 2. GLOBALS & PATHS ===

SCRIPT_DIR        = pathlib.Path(__file__).parent
WORK_DIR          = SCRIPT_DIR.parents[2]
HONOLULU_DATA_DIR = WORK_DIR / "00 - Data Sources" / "Honolulu"
TMK_LIST_PATH     = SCRIPT_DIR / "Target_Golf_Parcels_List.csv"
TAX_CSV_PATH      = HONOLULU_DATA_DIR / "All_Parcels_-4613852522541990741.csv"
OUT_CSV           = SCRIPT_DIR / "Phase5_Step5_Geographic_Breakdown.csv"

DISTRICT_MAP = {
    "1": "Honolulu (Urban Core)",
    "2": "Honolulu (East/Anomalies)",
    "3": "Honolulu (Anomalies)",
    "4": "Koolaupoko (Kailua/Kaneohe)",
    "5": "Koolauloa (North/East)",
    "6": "Waialua (North Shore)",
    "7": "Wahiawa (Central)",
    "8": "Waianae (West)",
    "9": "Ewa (Kapolei/Pearl City)",
}


# === 3. FUNCTIONS ===

# (none)


# === 4. EXECUTION ===

def main():
    print("\n" + "=" * 70)
    print("Phase 5 - Step 5: Geographic Concentration Breakdown")
    print("=" * 70 + "\n")

    if not TMK_LIST_PATH.exists():
        print(f"[FATAL] TMK list not found. Run Step 2 first.\n  {TMK_LIST_PATH}")
        raise SystemExit(1)
    if not TAX_CSV_PATH.exists():
        print(f"[FATAL] Cadastral CSV not found:\n  {TAX_CSV_PATH}")
        raise SystemExit(1)

    tmk_df = pd.read_csv(TMK_LIST_PATH)
    tmk_df.rename(columns={tmk_df.columns[0]: "TMK"}, inplace=True)
    tmk_df["TMK_clean"] = tmk_df["TMK"].astype(str).str.replace(r"[^0-9]", "", regex=True)

    tax_data = pd.read_csv(TAX_CSV_PATH)
    tmk_col  = next(
        (c for c in tax_data.columns if re.search(r"(?i)^tmk$", c)),
        None,
    )
    if tmk_col is None:
        print("[FATAL] No TMK column identified in cadastral CSV.")
        print("Available columns:", list(tax_data.columns))
        raise SystemExit(1)

    tax_data["TMK_clean"] = tax_data[tmk_col].astype(str).str.replace(r"[^0-9]", "", regex=True)

    step2_lens = tmk_df["TMK_clean"].str.len()
    csv_lens   = tax_data["TMK_clean"].dropna().str.len()

    # 8-digit format = Z S PPP QQQ  (3-digit parcel field)
    # 9-digit format = Z S PPP QQQQ (4-digit parcel field, trailing 0 for non-CPR parcels)
    if (step2_lens == 8).all() and (csv_lens == 9).all():
        tmk_df["TMK_clean"] = tmk_df["TMK_clean"] + "0"
    elif (step2_lens == 9).all() and (csv_lens == 8).all():
        tax_data["TMK_clean"] = tax_data["TMK_clean"] + "0"

    merged_data = tmk_df.merge(tax_data, on="TMK_clean", how="inner")

    merged_data["Zone_Code"] = merged_data["Zone"].astype(str)
    merged_data["District_Name"] = merged_data["Zone_Code"].map(DISTRICT_MAP).fillna(
        "Zone " + merged_data["Zone_Code"]
    )

    geo_summary = (
        merged_data
        .groupby(["Zone_Code", "District_Name"], as_index=False)
        .agg(Parcel_Count=("Zone_Code", "count"))
        .assign(Pct_of_Total_Parcels=lambda df: df["Parcel_Count"] / df["Parcel_Count"].sum() * 100)
        .sort_values("Parcel_Count", ascending=False)
        .reset_index(drop=True)
    )

    print(f"{'Zone':<5} {'Geographic District':<35} {'Parcel Count':<15} {'% of Parcels':<15}")
    print("-" * 70)
    for _, row in geo_summary.iterrows():
        print(
            f"{row['Zone_Code']:<5} {row['District_Name']:<35} "
            f"{int(row['Parcel_Count']):<15} {row['Pct_of_Total_Parcels']:.1f}%"
        )
    print("-" * 70)
    print(f"{'':5} {'TOTAL':<35} {int(geo_summary['Parcel_Count'].sum()):<15} 100.0%")
    print("-" * 70)

    geo_summary.to_csv(OUT_CSV, index=False)
    print(f"\n[+] Geographic Breakdown saved -> {OUT_CSV.name}")
    print("\n[DONE] Step 5 Complete. Phase 5 is fully finished!")


if __name__ == "__main__":
    main()
